File: Day 2/AOC2023_2_P2.rb
def extractVal(set)
    set.split(",").map do |x|
        [x.split(" ")[1], x.split(" ")[0].to_i]
    end.to_h
end

def processLine(line)
    tmp = { "red" => 0, "green" => 0, "blue" => 0}
    # Process each set in a game
    line.split(":")[-1].split(";").each do |x| 
        extractVal(x).each_pair do |k, v|
            if tmp[k] < v
                tmp[k] = v
            end
        end
    end
    return tmp.values.reduce(:*)
end

puts File.open("inputs").each.map {|x| processLine(x).to_i}.sum

File: Day 2/AOC2023_2_P2.nim
import std/tables
import std/strutils
import std/sugar
import std/math
import std/sequtils

proc processLine(line: string) : int = 
    var tmp = {"red": 0, "green": 0, "blue": 0}.toTable
    let infos = line.split(seps = {':', ',', ';'})[1 .. ^1]
    for elt in infos:
        let (v, k) = (elt.strip().split(" ")[0], elt.strip().split(" ")[1])
        if tmp[k] < parseInt(v):
            tmp[k] = parseInt(v)
    let val = collect:
        for elt in tmp.values:
            elt
    return foldl(val, a * b)

let res = collect:
    for elt in lines("inputs"):
        processLine(elt)
echo sum(res)